### PR TITLE
Add Element Hiding Support

### PR DESCRIPTION
**Pull request by dharb**
**Reviewer:** @jdorweiler 

**CC:** @jdorweiler @russellholt @kdzwinel @andrey-p 

## Description:
Hide iframes and their wrappers when they contain blocked requests. This adds functionality to deal with three common forms of ads:
1. Top level third-party iframe:
`<div class="ad-class">`	        <--- ad wrapper
	    `<iframe src="...">`		<--- blocked request
	    `</iframe>`
`</div>`

2. Nested third-party iframe
`<div class="ad-class">`	        <--- ad wrapper
	`<iframe src="...">`            <--- first party or otherwise unblocked request
		`<iframe src="...">`    <--- blocked request
		`</iframe>`
	`</iframe>`
`</div>`

3. Nested third-party scripts that would have loaded iframes
`<div class="ad-class">`	        <--- ad wrapper
	`<iframe src="...">`            <--- first party or otherwise unblocked request
		`<script src="...">`	<--- blocked request
                `</script>`
	`</iframe>`
`</div>`

I understand this code might be a bit hard to follow, since it relies on multiple content scripts running simultaneously in different iframes. See comments in element-hiding.js for more details. The rough description of what this does is:

1. Inject content scripts into all frames.
2. In each content script, wait for document to hit `readyState == 'interactive'`, which means the DOM in the frame is loaded (but scripts and other assets may not be yet).
3. a) If content script is running in main frame- gather list of top level iframes, store it, and send message to background page specifying that message is from main frame.
b) If content script is running in iframe - gather list of iframes and scripts contained within iframe, store it, and send message to background page specifying that message is from iframe.
4. In background page, listen for these messages from content scripts. If message is from main frame, respond with list of blocked requests of type frame. If message is from iframe, respond with list of blocked requests of type frame or type script.
5. In content scripts, listen for this message from background page and then:
a) If in main frame, iterate through the list of top level iframes, checking each src to see if it's included in the list of blocked frame requests. If it is, hide it and hide its parent node if the frame was the only child.
b) If in iframe, send message to content script in main frame containing list of blocked requests.
6. In main frame, listen for the message from 5b. When message is received, iterate through list of top level iframes and respond with the DOM id of each iframe element.
7. In each iframe, listen for message from 6 containing the frame's DOM id, then iterate through the list of found scripts and frames from 3b checking to see if any are in the list of blocked requests. If match is found, pass message to main frame containing the iframe's DOM id.
8. In main frame, listen for message containing frame id. When message is received, hide element with matching frame id and iterate upward recursively (if containers only contains iframe)

## Steps to test this PR:
1. Visit yahoo.com or other oath site where this is being tested
2. See that ads containing third party trackers are collapsed on page, but first party ads are not. finance.yahoo.com seems to consistently have first party ads that won't be hidden by this script. The recursive element hiding can also be tested on finance.yahoo.com -- the banner ad has a static container about five layers up from the iframe.
3. Visit random website, see that background page disables content scripts.

## Automated tests:
- [ ] Unit tests
- [ ] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

File: shared/js/background/redirect.es6.js [hunk 1]
```diff
@@ -10,6 +10,23 @@ var debugRequest = false
...
+    p.onMessage.addListener(function(m) {
+        if (m.frame === 'main') {
+            p.postMessage({blockedRequests: requestTab.framesBlocked});
+        } else if (m.frame === 'sub') {
+            p.postMessage({blockedRequests: requestTab.scriptsAndFramesBlocked});
+        }
+    })
+}
+
+chrome.runtime.onConnect.addListener(connected);
```
Comment (andrey-p) [inline shared/js/background/redirect.es6.js]: Can we move this to `chrome-events`?

Comment (dharb) [reply shared/js/background/redirect.es6.js]: @andrey-p definitely. I still need to refactor and organize the code quite a bit, but just wanted to get a working version up here to get more eyes on it.

Comment (andrey-p) [reply shared/js/background/redirect.es6.js]: Sorry, I'm jumping the gun a bit :gun: :man_cartwheeling:

Comment (dharb) [reply shared/js/background/redirect.es6.js]: No worries at all! I'm just happy you're looking at it already.

File: shared/js/background/redirect.es6.js [hunk 2]
```diff
@@ -10,6 +10,23 @@ var debugRequest = false
 
 trackers.loadLists()
 
+var ports = []
+function connected(p) {
+    ports[p.sender.frameId] = p;
```
Comment (jdorweiler) [inline shared/js/background/redirect.es6.js]: What is this used for?

Comment (dharb) [reply shared/js/background/redirect.es6.js]: mainly to establish an array of connections between the background page and each iframe so pipes don't get crossed.  I kinda don't like the idea of having a sparsely populated array there, so I might see if there's a better approach here.

Comment (jdorweiler) [reply shared/js/background/redirect.es6.js]: Where is it used though?

Comment (dharb) [reply shared/js/background/redirect.es6.js]: that's actually a good point. I had been using it in an earlier version to send messages back to the content script in the main frame of the active tab upon receiving messages from other content scripts on the page. I ended up not needing it once I started passing messages directly from iframes to the top frame.

I'll remove.

File: shared/js/content-scripts/element-hiding.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        if (e.data.frameId) {
+            frameId = e.data.frameId;
+            mainFrameUrl = e.data.mainFrameUrl
+            contentScript.locateBlockedFrames(e.data.blockedRequests);
+        }
+    }
+});
+
+csPort.onMessage.addListener((m) => {
+    if (!isTop) {
```
Comment (jdorweiler) [inline shared/js/content-scripts/element-hiding.js]: Can you switch this to `if (isTop) {} else {}` to stay consistent with the others.

File: shared/js/content-scripts/element-hiding.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+var frameId;
+var mainFrameUrl;
+var possibleTargets = [];
+var foundScripts = [];
+var foundFrames = [];
+
+
+window.addEventListener("message", (e) => {
+    if (isTop) {
+        if (e.data.frameUrl) {
```
Comment (jdorweiler) [inline shared/js/content-scripts/element-hiding.js]: Would it make sense to pass some data.type in the message instead of checking for frameUrl or frameId? `data.type === 'hide element'` That might make it easier to follow what each message is doing.

File: shared/js/content-scripts/element-hiding.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
+/**
+ * The purpose of this content script is to hide third-party tracking ads that have been blocked
+ * and appear as blank spaces on websites. In order to distinguish between tracking ads and legitimate
+ * ads, it is necessary to inject this script into all frames.
+ */
+
+var contentScript = {}
+contentScript.frameType = window === window.top ? 'main' : window.parent === window.top ? 'topLevelFrame' : 'nestedFrame'
```
Comment (dharb) [inline shared/js/content-scripts/element-hiding.js]: After seeing some deeply nested iframes (4+ layers), I decided to limit most of this to top level iframes for performance.

File: shared/js/content-scripts/element-hiding.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+ * 2. main frame sends messages back to iframes with their id (setFrameId)
+ * 3. iframes send messages to main frame when they contain blocked elements (hideFrame)
+ */
+contentScript.frameListener = (e) => {
+    if (e.data.type === 'frameIdRequest') {
+        contentScript.foundFrames = document.getElementsByTagName('iframe')
+        let i = contentScript.foundFrames.length
+        while (i--) {
+            let frame = contentScript.foundFrames[i]
+            if (frame.src && frame.id && !e.data.blockedRequests.includes(frame.src) && !frame.className.includes('ddg-hidden')) {
```
Comment (dharb) [inline shared/js/content-scripts/element-hiding.js]: Can't inject content scripts into blocked iframes, so no point messaging them. 

Also no point messaging iframes that have already been hidden. My first thought was to add a class to every frame that had already been messaged like this, and not message them again. Unfortunately, some ad scripts dynamically change the src url of iframes, which means this has to be run again.

File: shared/js/content-scripts/element-hiding.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+ * Hide frames that were either themselves blocked, or that contain scripts
+ * or other frames that were blocked. Then traverse DOM upward, hiding
+ * parent selector if it only contains the blocked frame. Add class
+ * and remove event listeners so that other content scripts no longer
+ * interact with hidden frames
+ */
+contentScript.hideFrame = (frame) => {
+    frame.style.setProperty('display', 'none', 'important')
+    frame.hidden = true
+    frame.className += ' ' + 'ddg-hidden'
```
Comment (kdzwinel) [inline shared/js/content-scripts/element-hiding.js]: nit: frame.classList.add('ddg-hidden')

File: shared/js/content-scripts/element-hiding.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+/**
+ * Set up listener in each frame to respond to messages from other frames.
+ * There are three types of messages being passed:
+ * 1. iframes send messages to the main frame requesting their id (frameIdRequest)
+ * 2. main frame sends messages back to iframes with their id (setFrameId)
+ * 3. iframes send messages to main frame when they contain blocked elements (hideFrame)
+ */
+contentScript.frameListener = (e) => {
+    if (e.data.type === 'frameIdRequest') {
+        contentScript.foundFrames = document.getElementsByTagName('iframe')
```
Comment (kdzwinel) [inline shared/js/content-scripts/element-hiding.js]: I think `<iframe>` is not the only "frame object" - there is also `<frame>`, `<object>` and `<embed>` that (as far as I remember) can be used similarly. We can get all of those with `window.frames`.

File: shared/js/background/chrome-events.es6.js [hunk 9]
```diff
@@ -124,6 +124,22 @@ chrome.runtime.onMessage.addListener((req, sender, res) => {
         ATB.setAtbValuesFromSuccessPage(req.atb)
     }
 
+    // listen for messages from content scripts injected into frames
+    // on specific domains. Respond with list of blocked requests.
+    if (req.hideElements) {
+        let requestTab = tabManager.get({tabId: sender.tab.id})
+        if (requestTab.parentEntity === 'Oath') {
+            if (req.frame === 'main') {
+                chrome.tabs.sendMessage(sender.tab.id, {type: 'blockedRequests', blockedRequests: requestTab.framesBlocked, frame: 'main'}, {frameId: sender.frameId})
```
[the remark this answers is not in the document]
Comment (jdorweiler) [reply shared/js/background/chrome-events.es6.js]: This might be why I'm sometimes seeing blocked ads that are not hidden.

File: shared/js/content-scripts/element-hiding.js [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        if (req.frame === 'main') {
+            contentScript.locateBlockedFrames(req.blockedRequests)
+        } else if (req.frame === 'topLevelFrame') {
+            contentScript.foundScripts = document.getElementsByTagName('script')
+            contentScript.possibleTargets = Array.from(contentScript.foundFrames).concat(Array.from(contentScript.foundScripts))
+            window.top.postMessage({frameUrl: document.location.href, blockedRequests: req.blockedRequests, type: 'frameIdRequest'}, req.mainFrameUrl)
+        }
+    } else if (req.type === 'disable') {
+        window.removeEventListener('message', contentScript.frameListener)
```
Comment (jdorweiler) [inline shared/js/content-scripts/element-hiding.js]: 👍

File: shared/js/background/chrome-events.es6.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (dharb) [reply shared/js/background/chrome-events.es6.js]: This is a good idea. I should be able to switch things around here.

File: shared/js/background/chrome-events.es6.js [hunk 11]
```diff
@@ -124,6 +124,22 @@ chrome.runtime.onMessage.addListener((req, sender, res) => {
         ATB.setAtbValuesFromSuccessPage(req.atb)
     }
 
+    // listen for messages from content scripts injected into frames
+    // on specific domains. Respond with list of blocked requests.
+    if (req.hideElements) {
+        let requestTab = tabManager.get({tabId: sender.tab.id})
+        if (requestTab.parentEntity === 'Oath') {
```
[the remark this answers is not in the document]
Comment (dharb) [reply shared/js/background/chrome-events.es6.js]: Oath is the parent company of yahoo. This initial roll out is limited to their properties, as we've seen a lot of complaints about them specifically. We'll expand to other sites in a future update once this has been vetted.

File: shared/js/background/classes/tab.es6.js [hunk 12]
```diff
@@ -34,9 +35,12 @@ class Tab {
         this.id = tabData.id || tabData.tabId
         this.trackers = {}
         this.trackersBlocked = {}
+        this.framesBlocked = ''
+        this.scriptsAndFramesBlocked = ''
```
[the remark this answers is not in the document]
Comment (dharb) [reply shared/js/background/classes/tab.es6.js]: Good call, I like that.

File: shared/js/content-scripts/element-hiding.js [hunk 13]
```diff
@@ -0,0 +1,122 @@
+/**
+ * The purpose of this content script is to hide third-party tracking ads that have been blocked
+ * and appear as blank spaces on websites. In order to distinguish between tracking ads and legitimate
+ * ads, it is necessary to inject this script into all frames.
+ */
+
+var contentScript = {}
```
[the remark this answers is not in the document]
Comment (dharb) [reply shared/js/content-scripts/element-hiding.js]: Good point. I wasn't totally sure about safari's level of support for some ES6 features, but I just did some digging and it looks like the versions of safari where the extension may be installed (10+) supports everything we need.

File: shared/js/content-scripts/element-hiding.js [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    if (e.data.type === 'frameIdRequest') {
+        contentScript.foundFrames = document.getElementsByTagName('iframe')
+        let i = contentScript.foundFrames.length
+        while (i--) {
+            let frame = contentScript.foundFrames[i]
+            if (frame.src && frame.id && !e.data.blockedRequests.includes(frame.src) && !frame.className.includes('ddg-hidden')) {
+                frame.contentWindow.postMessage({frameId: frame.id, mainFrameUrl: document.location.href, blockedRequests: e.data.blockedRequests, type: 'setFrameId'}, '*')
+            }
+        }
+    } else if (e.data.type === 'hideFrame') {
```
[the remark this answers is not in the document]
Comment (dharb) [reply shared/js/content-scripts/element-hiding.js]: Probably a good call. We can generate a random token identifier in the background page and send it to the content scripts in all frames, then check that it matches here before hiding anything.

File: shared/js/content-scripts/element-hiding.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (dharb) [reply shared/js/content-scripts/element-hiding.js]: For timeliness, I'll address this in another PR.

File: shared/js/content-scripts/element-hiding.js (hunk 8 above)
Comment (dharb) [reply shared/js/content-scripts/element-hiding.js]: I think for now, just for the sake of timeliness, we only care about iframes. Good catch though, we can address in a future PR.

File: shared/js/background/chrome-events.es6.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (dharb) [reply shared/js/background/chrome-events.es6.js]: Since there's some time pressure to get this in the upcoming release, I'll address this in a future PR. I agree it would improve the approach, but for now it's 'good enough'.

**Comment by jdorweiler** [comment]
👍 Looks great!

**Comment by dharb** [comment]
This is ready for beta testing. There are still some things I'd like to polish up, like adding a few more tests and addressing a few of @kdzwinel's suggestions (thanks for the thorough review!), but they shouldn't change the functionality significantly. I'll make those updates in subsequent PRs.
